sprint_8-solution-time-and-duration

itsmekatrine opened the pull request:
Доработана функциональность по техническому заданию спринта

Commit:
Сделать историю посещений неограниченной по размеру
Добавить метод void remove(int id) для удаления задачи из просмотра
Разработка алгоритма со связным списком и HashMap

Commit:
Добавить unit-тесты для:
- встроенного связного списка версий, а также операций добавления и удаления
- удаляемых подзадач, они не должны хранить внутри себя старые id
- эпиков, внутри них не должно оставаться неактуальных id подзадач
- неизменяемости полей в задаче

Commit:
Исправить CheckStyle для PR

Commit:
Исправить CheckStyle для PR

Commit:
Исправить CheckStyle для PR

Commit:
Исправить CheckStyle для PR

Commit:
Исправить CheckStyle для PR

Commit:
Сделать поля private

Commit:
В TaskTest для инициализации использовать класс-фабрику Managers

Commit:
Исправить двойное присваивание

Commit:
- Cкладывать в мапу текущую добавляемую задачу, а не хвост списка
- В метод linkLast передавать уже готовый объект ноды
- Удалить старую ноду из списка

Commit:
Поменять название map на history

Commit:
При удалении задач добавить удаление их из истории

Commit:
Добавлена техника "пустая голова и пустой хвост"

Commit:
Скорректировать юнит-тесты
Добавить новые юнит-тесты

Commit:
Исправить CheckStyle для PR

Commit:
Исправить CheckStyle для PR

Commit:
Исправить CheckStyle для PR

Commit:
Исправить CheckStyle для PR

Commit:
Исправить CheckStyle для PR

Commit:
Исправить CheckStyle для PR

Commit:
Исправить CheckStyle для PR

Commit:
Исправить CheckStyle для PR

Commit:
Убрать проверки на пустоту в методе linkLast()

Commit:
Убрать проверки на пустоту в методе removeNode()

Commit:
- Создать класс TaskType
- Добавить тип задачи в конструкторы задачи, сабтаски и эпика
- Реализовать класс FileBackedTaskManager

Commit:
Покрыть тестами класс FileBackedTaskManager

Commit:
Удаление ненужных импортов

Commit:
Исправление checkstyle

Commit:
Удаление flush после каждого типа задач

Commit:
Добавить переопределение методов удаления задач

Commit:
Отказ от instanceof, когда есть иерархия классов

Commit:
Переопределить метод toString в классе Task

Commit:
Вынести класс ManagerSaveException в отдельный пакет exception

Commit:
Public static TasksManager getDefault() реализацию InMemoryTasksManager нужно заменить на FileBackedTasksManager
+ правка тестов

Commit:
Исправление build tests

Commit:
Исправление build tests

Commit:
Добавить продолжительность и дату старта

Commit:
Добавить юнит-тесты на проверку новых полей

Commit:
Добавить вывод списка задач в порядке приоритета
+ поправить другие юнит-тесты

Commit:
Проверить пересечения с помощью Stream API

Commit:
Поменять циклы на Stream API

Commit:
Покрыть код тестами
- добавить расчёт статуса Epic
- добавить abstract class TaskManagerTest<T extends TaskManager>
- добавить тесты для HistoryManager
- тесты для проверки перехвата исключений

Commit:
Исправления для checkstyle

Yang-Pi commented on src/com/yandex/app/model/Subtask.java:
```diff
@@ -1,18 +1,31 @@
+
 public class Subtask extends Task {
     private int epicId;
+    private final TaskType type;
```
не нужно дублировать эти поля - они и так наследуются от Task и доступ к ним происходит от также унаследованных геттеров / сеттеров

itsmekatrine replied:
Исправлено в коммите b6c28cb1e679b952b722d4489edb348ed1ea8134

Yang-Pi commented on src/com/yandex/app/model/Epic.java:
```diff
@@ -1,22 +1,46 @@
 
 public class Epic extends Task {
     private List<Subtask> subtasks;
+    private final TaskType type;
```
не нужно дублировать эти поля - они и так наследуются от Task и доступ к ним происходит от также унаследованных геттеров / сеттеров

itsmekatrine replied:
Исправлено в коммите b6c28cb1e679b952b722d4489edb348ed1ea8134

Yang-Pi commented on src/com/yandex/app/model/Epic.java:
```diff
@@ -1,22 +1,46 @@
+    private final TaskType type;
+    private Duration duration;
+    private LocalDateTime startTime;
+    private LocalDateTime endTime;
```
достаточно оставить в эпике только endTime

itsmekatrine replied:
Исправлено в коммите b6c28cb1e679b952b722d4489edb348ed1ea8134

Yang-Pi commented on src/com/yandex/app/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,294 @@
+import java.util.List;
+import java.util.TreeSet;
+
+public class FileBackedTaskManager extends InMemoryTaskManager {
```
этот класс нужно слегка изменить с точки зрения записи в файл и считывания из файла - добавить обработку новых полей, в остальном вся логика должна быть в InMemoryTaskManager - prioritizedTasks, рефакторинг методов create, update, delete - всё там

itsmekatrine replied:
Исправлено в коммите 2c3edcde624bb176a034a23ae3b4b7058e781020

Yang-Pi commented on src/com/yandex/app/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,294 @@
+    public FileBackedTaskManager(File file, HistoryManager history) {
+        super(history);
+        this.file = file;
+        prioritizedTasks = new TreeSet<>((task1, task2) -> {
```
есть метод Comparator.comparing, который делает ровно то, что описано в длинной лямбде ниже, но в одну строчку - компактнее использовать его, передав в его аргумент ссылку на метод getStartTime

itsmekatrine replied:
Исправлено в коммите d98245199d0e1be7db7d88c585ea281dc7bf46eb

Yang-Pi commented on src/com/yandex/app/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,294 @@
+        if (oldSubtask != null) {
+            prioritizedTasks.remove(oldSubtask);
+
+            boolean hasCrossTasks = getAllSubtasks().stream()
```
логика с нахождением пересечений дублируется, поэтому ее лучше вынести в отдельный метод и переиспользовать

itsmekatrine replied:
Исправлено в коммите 98f625625718508e23ed26f20a524c2cab9fed44

Yang-Pi commented on src/com/yandex/app/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,294 @@
+        boolean result = super.deleteTaskById(id);
+        if (result) {
+            Task task = getTaskById(id);
+            prioritizedTasks.remove(task);
```
молодец, что не забываешь удалять также из этого множества

Commit:
Исправить дублирование нахождения пересечений

Commit:
Исправить в списке сортировки лямбду на метод Comparator.comparing

Commit:
- Перенести логику в InMemoryTaskManager - prioritizedTasks, рефакторинг методов create, update, delete
- Исправить запись в файл и считывание из файла - добавить обработку новых полей

Commit:
Удалить дублирующиеся поля

Commit:
Исправления checkstyle

Yang-Pi commented on src/com/yandex/app/service/InMemoryTaskManager.java:
```diff
@@ -132,56 +167,64 @@ public Subtask getSubtaskById(int id) {
+                epic.updateEpicStatus();
             }
         }
+        save();
```
а зачем во многих методах вызывается пустой метод save?
в FileBackedTaskManager же есть переопределенные методы, которые кроме вызова метода родителя дополнительно вызывают логику save() - там это логично, а в InMemoryTaskManager метод save не нужен концептуально

itsmekatrine replied:
Исправлено в коммите 2b54db9239bd29686463fb6405a2107072282edd

Yang-Pi commented on src/com/yandex/app/model/Epic.java:
```diff
@@ -34,25 +52,57 @@ public boolean hasSubtask(Epic epic, Subtask subtask) {
+        this.endTime = LocalDateTime.MIN;
+
+        for (Task subtask : subtasks) {
+            this.duration = this.duration.plus(subtask.getDuration());
```
необязательно: продолжительность можно посчитать тоже через Stream API, используя map и reduce операции

itsmekatrine replied:
Исправлено в коммите 3630a795c303a722a62fa27ea950bc5f27ed92c5

Yang-Pi commented on src/com/yandex/app/model/Epic.java:
```diff
@@ -34,25 +52,57 @@ public boolean hasSubtask(Epic epic, Subtask subtask) {
+            if (subtask.getStartTime() != null && subtask.getStartTime().isBefore(this.startTime)) {
+                this.startTime = subtask.getStartTime();
+            }
+            if (subtask.getEndTime() != null && subtask.getEndTime().isAfter(this.endTime)) {
```
необязательно: время начала и конца можно посчитать тоже через Stream API, используя map и min/max операции (Comparator.naturalOrder())

itsmekatrine replied:
Исправлено в коммите 3630a795c303a722a62fa27ea950bc5f27ed92c5

Yang-Pi commented on src/com/yandex/app/model/Epic.java:
```diff
@@ -34,25 +52,57 @@ public boolean hasSubtask(Epic epic, Subtask subtask) {
+    }
+
+    // подсчёт продолжительности выполнения эпика
+    public Duration calculateEpicDuration() {
```
еще тут считается startTime и endTime

itsmekatrine replied:
Исправлено в коммите 3630a795c303a722a62fa27ea950bc5f27ed92c5

Yang-Pi commented on src/com/yandex/app/service/InMemoryTaskManager.java:
```diff
@@ -26,69 +39,82 @@ public List<Task> getAllTasks() {
-        for (int id : idOfTasks) {
-            deleteTaskById(id);
-        }
+        tasks.keySet().forEach(this::deleteTaskById);
```
здесь тоже нужно удалять из prioritizedTasks

itsmekatrine replied:
Исправлено в коммите 249ea593dddae93207058e4e11942d219c016134

Yang-Pi commented on src/com/yandex/app/service/InMemoryTaskManager.java:
```diff
@@ -132,56 +167,64 @@ public Subtask getSubtaskById(int id) {
-                    epic.updateEpicStatus();
-                }
-            }
+            history.remove(id);
```
здесь тоже нужно удалять из prioritizedTasks

itsmekatrine replied:
Исправлено в коммите 249ea593dddae93207058e4e11942d219c016134

Yang-Pi commented on src/com/yandex/app/service/InMemoryTaskManager.java:
```diff
@@ -132,56 +167,64 @@ public Subtask getSubtaskById(int id) {
-        for (int id : idOfSubtasks) {
-            deleteSubtaskFromEpic(id);
-        }
+        subtasks.keySet().forEach(this::deleteSubtaskFromEpic);
```
здесь тоже нужно удалять из prioritizedTasks

itsmekatrine replied:
Исправлено в коммите 249ea593dddae93207058e4e11942d219c016134

Yang-Pi commented on src/com/yandex/app/service/InMemoryTaskManager.java:
```diff
@@ -214,33 +259,66 @@ public void updateEpic(int id, Epic epic) {
     @Override
     public void deleteAllEpics() {
-        List<Integer> idOfEpics = new ArrayList<>(epics.keySet());
+        epics.keySet().forEach(this::deleteEpicById);
```
здесь тоже нужно удалять из prioritizedTasks

itsmekatrine replied:
Исправлено в коммите 249ea593dddae93207058e4e11942d219c016134

Yang-Pi commented on src/com/yandex/app/service/InMemoryTaskManager.java:
```diff
@@ -214,33 +259,66 @@ public void updateEpic(int id, Epic epic) {
+        epics.keySet().forEach(this::deleteEpicById);
+    }
+
+    protected boolean hasCrossingTasks(Task task) {
```
почему не private? метод используется только в этом классе

itsmekatrine replied:
Исправлено в коммите 3b468b784afc87f16866f695b0b8f55b3aad9a30

Yang-Pi commented on src/com/yandex/app/service/InMemoryTaskManager.java:
```diff
@@ -214,33 +259,66 @@ public void updateEpic(int id, Epic epic) {
+                .anyMatch(existingTask -> isCrossTasks(existingTask, task));
+    }
+
+    public boolean isCrossTasks(Task task1, Task task2) {
```
почему не private? метод используется только в этом классе

itsmekatrine replied:
Исправлено в коммите 3b468b784afc87f16866f695b0b8f55b3aad9a30

Yang-Pi commented on src/com/yandex/app/service/InMemoryTaskManager.java:
```diff
@@ -214,33 +259,66 @@ public void updateEpic(int id, Epic epic) {
+        return start1.isBefore(end2) && start2.isBefore(end1);
+    }
+
+    public void checkCrossTasks() {
```
этот метод не используется

itsmekatrine replied:
Исправлено в коммите bae98a1725682e6a73424cac6b3b7764f7bdea6b

Yang-Pi reviewed (commented):
хорошая работа! оставил несколько замечаний и рекомендаций)

Commit:
Удаление неиспользующегося метода

Commit:
Исправить видимость методов

Commit:
Добавить удаление из prioritizedTasks

Commit:
Убрать метод save()

Commit:
Исправить подсчёт продолжительности, времени начала и конца через Stream API
